pkg/globalnet/controllers/cluster_egressip_controller.go

```diff
@@ -128,14 +145,177 @@ func (c *clusterGlobalEgressIPController) onCreate(egressIP *submarinerv1.Cluste
+	}
+
+	if *numberOfIPs == 0 {
+		tryAppendStatusCondition(&egressIP.Status, &metav1.Condition{
+			Type:    string(submarinerv1.GlobalEgressIPAllocated),
+			Status:  metav1.ConditionFalse,
+			Reason:  "ZeroInput",
+			Message: "No global IPs to allocate",
+		})
+
+		return errors.Errorf("NumberOfIPs %q in %q cannot be 0", *numberOfIPs, egressIP.Name)
+	}
+
+	return nil
+}
+
+func (c *clusterGlobalEgressIPController) onCreate(key string, numberOfIPs *int, status *submarinerv1.GlobalEgressIPStatus) bool {
+	// When Globalnet Pod is restarted or migrated, we just have to Sync the IPPool cache and IPTable rules.
+	if *numberOfIPs == len(status.AllocatedIPs) {
+		err := c.pool.Reserve(status.AllocatedIPs...)
+		if err != nil {
+			klog.Errorf("Error allocating IPs %v for %q: %v", status.AllocatedIPs, key, err)
+			tryAppendStatusCondition(status, &metav1.Condition{
+				Type:    string(submarinerv1.GlobalEgressIPAllocated),
+				Status:  metav1.ConditionFalse,
+				Reason:  "IPPoolAllocationFailed",
+				Message: fmt.Sprintf("Error allocating %v global IP(s) from the pool: %v", status.AllocatedIPs, err),
+			})
+
+			return false
+		}
+
+		if err := c.programClusterGlobalEgressRules(status); err != nil {
+			_ = c.pool.Release(status.AllocatedIPs...)
+			status.AllocatedIPs = []string{}
+
+			klog.Errorf("Error syncing the IPTable rules on the node for %q: %v", key, err)
+
+			return true
+		}
+
 		return false
 	}
 
-	return allocateIPs(key, egressIP.Spec.NumberOfIPs, c.pool, &egressIP.Status)
+	return c.allocateGlobalIPs(key, numberOfIPs, status)
 }
 
-func (c *clusterGlobalEgressIPController) onRemove(egressIP *submarinerv1.ClusterGlobalEgressIP) bool { // nolint unparam
-	// TODO - remove IP table rules for the allocated IPs
+func (c *clusterGlobalEgressIPController) onUpdate(key string, numberOfIPs *int, status *submarinerv1.GlobalEgressIPStatus) bool {
+	if *numberOfIPs == len(status.AllocatedIPs) {
+		klog.V(log.DEBUG).Infof("Update called for %q, but numberOfIPs %q are already allocated", key, *numberOfIPs)
+		return false
+	}
+
+	// If numGlobalIPs is modified, delete the existing allocation.
+	if len(status.AllocatedIPs) > 0 {
+		c.flushClusterGlobalEgressRules(status)
+		_ = c.pool.Release(status.AllocatedIPs...)
+	}
+
+	return c.allocateGlobalIPs(key, numberOfIPs, status)
+}
+
+func (c *clusterGlobalEgressIPController) onRemove(key string, egressIP *submarinerv1.ClusterGlobalEgressIP) bool { // nolint unparam
+	if len(egressIP.Status.AllocatedIPs) > 0 {
+		c.flushClusterGlobalEgressRules(&egressIP.Status)
+		_ = c.pool.Release(egressIP.Status.AllocatedIPs...)
+	}
+
+	return false
+}
+
+func (c *clusterGlobalEgressIPController) flushClusterGlobalEgressRules(status *submarinerv1.GlobalEgressIPStatus) {
+	snatIP := c.getTargetSNATIPaddress(status.AllocatedIPs)
+	if snatIP == "" {
+		klog.Warning("flushClusterGlobalEgressRules called with 0 AllocatedIPs")
+		return
+	}
```

Comment:
I think it's reasonable to assume this is only called with `AllocatedIPs` > 0.  This would be a code path that is never hit. Same on L252.